File: 2882-ways-to-express-an-integer-as-sum-of-powers/2882-ways-to-express-an-integer-as-sum-of-powers.js
/**
 * @param {number} n
 * @param {number} x
 * @return {number}
 */
var numberOfWays = function(n, x) {
    const powers = [];
    let i = 1;
    while (true) {
        const pow = i ** x;
        if (pow > n) break;
        powers.push(pow);
        i++;
    }
    const dp = Array.from({length: n + 1}).fill(0);
    dp[0] = 1;
    for (const p of powers) {
        for (let i = n; i > p - 1; i--) {
            dp[i] = (dp[i] + dp[i - p]) % (10**9 + 7);
        }
    }
    return dp[n];
};

File: 2882-ways-to-express-an-integer-as-sum-of-powers/2882-ways-to-express-an-integer-as-sum-of-powers.py
class Solution:
    def numberOfWays(self, n: int, x: int) -> int:
        powers = []
        i = 1
        while True:
            pow = i ** x
            if pow > n:
                break
            powers.append(pow)
            i += 1
        dp = [0] * (n + 1)
        dp[0] = 1
        for p in powers:
            for s in range(n, p - 1, -1):
                dp[s] = (dp[s] + dp[s - p]) % (10**9 + 7)
        return dp[n]